framework/bootstrap/configcontroller.go

```diff
@@ -16,145 +21,109 @@ import (
 	"slime.io/slime/framework/bootstrap/viewstore"
 	"strconv"
 	"strings"
-
-	meshconfig "istio.io/api/mesh/v1alpha1"
-	mcpresource "istio.io/istio-mcp/pkg/config/schema/resource"
-	mcpc "istio.io/istio-mcp/pkg/mcp/client"
-	xdsc "istio.io/istio-mcp/pkg/mcp/xds/client"
-	mcpmodel "istio.io/istio-mcp/pkg/model"
+	"time"
 )
 
 const (
 	KubernetesConfigSourcePrefix = "k8s://"
 	McpOverXdsConfigSourcePrefix = "xds://"
 )
 
+type InitReady interface {
+	InitReady() bool
+}
+
 type ConfigController interface {
 	viewstore.ViewerStore
 	RegisterEventHandler(kind resource.GroupVersionKind, handler func(resource.Config, resource.Config, Event))
 	Run(stop <-chan struct{})
+	InitReady
 }
 
 type configController struct {
 	viewerStore        viewstore.ViewerStore
 	monitorControllers []*monitorController
-	stop               chan<- struct{}
+	stop               <-chan struct{}
 }
 
-func NewIstioConfigController(config *bootconfig.Config) (ConfigController, error) {
-	// XXX merge with NewIstioConfigController
-
-	configSource := config.Global.IstioConfigSource
-	if configSource == nil {
-		return nil, nil
+func RunIstioController(c ConfigController, config *bootconfig.Config) (ConfigController, error) {
+	cc, ok := c.(*configController)
+	if !ok {
+		return c, fmt.Errorf("convert to *configController error")
 	}
-
-	var cc *configController
-	stopCh := make(chan struct{})
-	var mcs []*monitorController
-
 	xdsSourceEnableIncPush := config.Global.Misc["xdsSourceEnableIncPush"] == "true"
+	stop := cc.stop
 
-	// init all monitorControllers
-	imc, err := initIstioMonitorController()
-	if err != nil {
-		return nil, err
-	} else {
-		mcs = append(mcs, imc)
-	}
-
-	if strings.HasPrefix(configSource.Address, McpOverXdsConfigSourcePrefix) {
-		if mc, err := initXdsMonitorController(configSource); err != nil {
-			return nil, err
-		} else {
-			mcs = append(mcs, mc)
-		}
-	}
-
-	vs, err := makeViewerStore(mcs)
-	if err != nil {
-		return nil, err
-	}
-
-	cc = &configController{
-		viewerStore:        vs,
-		monitorControllers: mcs,
-		stop:               stopCh,
-	}
-
-	go cc.Run(stopCh)
-
-	for _, mc := range mcs {
+	go cc.Run(stop)
+	for _, mc := range cc.monitorControllers {
 		switch mc.kind {
 		case McpOverXds:
-			if err = startXdsMonitorController(mc, "", xdsSourceEnableIncPush); err != nil {
+			if err := startXdsMonitorController(mc, "", xdsSourceEnableIncPush, stop); err != nil {
 				return nil, err
 			}
 		}
 	}
-
 	log.Infof("start IstioConfigController successfully")
-
 	return cc, nil
 }
 
-func NewConfigController(config *bootconfig.Config, cfg *rest.Config) (ConfigController, error) {
-	var cc *configController
-	stopCh := make(chan struct{})
+func NewConfigController(configSources []*bootconfig.ConfigSource, stop <-chan struct{}) (*configController, error) {
```

Comment:
I can't get why we change this.

Reply:
split into two parts: NewConfigController and RunController

prepare data struct in NewConfigController

run controller in RunController